services/app-api/src/postgres/contractAndRates/withdrawContract.ts

```diff
@@ -176,42 +194,78 @@ const withdrawContractInsideTransaction = async (
                         a.contractRevision.createdAt.getTime()
                 )
 
-        // get the parent contract data of this rate.
-        const parentContract = submissionPackages?.find(
-            (pkg) => pkg.contractRevision.contract.id === parentContractID
-        )?.contractRevision.contract
-
-        // check if any of the linked submissions are approved or submitted and not withdrawn
-        const hasApprovedOrSubmittedSubs = submissionPackages.some((pkg) => {
-            const reviewStatusAction =
-                pkg.contractRevision.contract.reviewStatusActions[0]?.actionType
-            const isSubmitted =
-                pkg.contractRevision.contract.revisions[0].submitInfo
-
-            // skip if this is the parent contract submission package.
-            if (pkg.contractRevision.contract.id === contract.id) return false
-            // is contract approved
-            if (reviewStatusAction === 'MARK_AS_APPROVED') return true
-            // is contract submitted and review status is not withdrawn
-            if (isSubmitted && reviewStatusAction !== 'WITHDRAW') return true
-            // otherwise does not meet condition
-            return false
-        })
-
         // skipping rate if conditions are met
         if (
             !latestRateRevision.submitInfo || // unlocked rate. linked rates can be in unlocked status
             rate.reviewStatusActions[0]?.actionType === 'WITHDRAW' || // is the rate already withdrawn
-            hasApprovedOrSubmittedSubs || // any of the linked submissions are approved or submitted
-            !parentContract // standalone rate could happen from a bug, old feature, or new feature. We don't want to withdraw these in this fashion.
+            parentContractID !== contract.id // if it is a linked rate
         ) {
             continue
         }
 
-        // TODO: Pending decision if rate can be withdrawn if its linked to a unlocked submission that had this rate in its latest submission package.
-        //  - If we do allow it, then delete this TODO, if we do not allow it we need to add logic to check unlocked
-        //      submissions for rate in its latest submission package.
-        ratesToWithdraw.push(rate)
+        let reassignToContractID:
+            | {
+                  contractID: string
+                  status: ConsolidatedContractStatus
+              }
+            | undefined = undefined
+
+        // Looping through each rate's latest revision's related submissions to find a valid new parent contract.
```

Comment:
suggestion:  think this long loop is a good candidate also for standalone function. Seems like it has has a type of something  like `prepareToWithdrawRates(rates) => [ratesToWithdraw, ratesToResassignParent]`